### PR TITLE
sasquatch: 4.4 -> 4.5.1-1

### PR DESCRIPTION
**Pull request by vlaci**
Required for #217836

###### Description of changes

Changed using ONEKEY [fork](https://github.com/onekey-sec/sasquatch) of sasquatch which is up-to-date with squashfs-tools 4.5.1 and the Darwin compatibility patch shipped in Nixpkgs applies cleanly as well.

I have chosen to override the `squashfsTools` derivation, as this fork is entirely compatible with that version. I am open to critique on this. I wasn't sure copying from squashfstools or lifting a `generic.nix` base derivation out of it is the better solution. Existing NixOS test is also nice to have as a validation, although it may be overkill here.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by vlaci** [comment]
Ah, figured keeping the existing test doesn't make sense as it will just use squashfs-tools anyway.

**Comment by AkechiShiro** [comment]
Any news, any blockage ? Or any review from @Pamplemousse if he is still the right contact for this PR ?

**Comment by vlaci** [comment]
Some questions got answerred: I shouldn't depend on squashFsTools at all, as it got updated to 4.6.1, and it is unlikely I want to rebase sasquatch the near future.

**Comment by vlaci** [comment]
As `squashfsTools` derivation is now incompatible with previous versions, I have recreated the derivation from scratch.

**Comment by Pamplemousse** [comment]
@AkechiShiro Sorry I missed your message earlier.

I have stopped using this package in years... I am not well placed to test.
Also, could you remove me from the maintainers please?

**Comment by vlaci** [comment]
> @AkechiShiro Sorry I missed your message earlier.
> 
> I have stopped using this package in years... I am not well placed to test. Also, could you remove me from the maintainers please?

Done

**Comment by qkaiser** [comment]
This PR has been stuck for the last 6 months waiting for a review from someone who explicitly requested to be removed from the maintainers list.

Who should review this then ? How can we make it move forward ?

Asking cause it's the last item missing in out meta-issue to get unblob in nixpkgs  https://github.com/NixOS/nixpkgs/issues/217836

**Comment by AkechiShiro** [comment]
This PR needs a rebase, first of all on master @vlaci could you please rebase please ?

**Comment by vlaci** [comment]
Happy new year everyone :)
Rebased on master.

**Comment by AkechiShiro** [comment]
Thanks @vlaci, there seems to be a typo in the meta attribute of the package, please fix it, so that the ofborg evaluation test passes, see more details : https://gist.github.com/GrahamcOfBorg/6d11645476eaaef9b7fc87c573d4a188, `meta.mainprogram` should be `meta.mainProgram`. 

I cannot help further as I'm inexperienced in reviews @qkaiser, maybe mentioning the pull request as ready for review here, should bring this PR to more experienced reviewers : https://discourse.nixos.org/t/prs-ready-for-review/3032 and hopefully help.

**Comment by qkaiser** [comment]
@AkechiShiro thanks for the heads up, I sent a reply to the Discourse thread asking for a review on this PR

**Review by tnias — APPROVED** [review]
I ran nixpkgs-review and sucessfully used both x86_64 and aarch64 to extract an example squashfs I found somewhere on my disk.
For me the diffoscope tests are broken on master on x86_64 even without this patch.

---

Result of `nixpkgs-review pr 219169` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages failed to build:</summary>
  <ul>
    <li>diffoscope</li>
    <li>diffoscope.dist</li>
    <li>diffoscope.man</li>
  </ul>
</details>
<details>
  <summary>10 packages built:</summary>
  <ul>
    <li>binwalk (python311Packages.binwalk)</li>
    <li>binwalk.dist (python311Packages.binwalk.dist)</li>
    <li>python310Packages.binwalk</li>
    <li>python310Packages.binwalk-full</li>
    <li>python310Packages.binwalk-full.dist</li>
    <li>python310Packages.binwalk.dist</li>
    <li>python311Packages.binwalk-full</li>
    <li>python311Packages.binwalk-full.dist</li>
    <li>sasquatch</li>
    <li>sasquatch-v4be</li>
  </ul>
</details>

Result of `nixpkgs-review pr 219169` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>13 packages built:</summary>
  <ul>
    <li>binwalk (python311Packages.binwalk)</li>
    <li>binwalk.dist (python311Packages.binwalk.dist)</li>
    <li>diffoscope</li>
    <li>diffoscope.dist</li>
    <li>diffoscope.man</li>
    <li>python310Packages.binwalk</li>
    <li>python310Packages.binwalk-full</li>
    <li>python310Packages.binwalk-full.dist</li>
    <li>python310Packages.binwalk.dist</li>
    <li>python311Packages.binwalk-full</li>
    <li>python311Packages.binwalk-full.dist</li>
    <li>sasquatch</li>
    <li>sasquatch-v4be</li>
  </ul>
</details>

**Comment by Mic92** [comment]
What does the bigendian variant do? Do we need both?

File: pkgs/tools/filesystems/sasquatch/default.nix
Comment (tilpner) [inline pkgs/tools/filesystems/sasquatch/default.nix:15]: This repeats the version number (v4), that's already appended to pname, so the full name would be `sasquatch-v4be-4.5.1.-4`. Perhaps it should just be `sasquatch-be`?

**Comment by qkaiser** [comment]
> What does the bigendian variant do? Do we need both?

sasquatch is built on top of squashfs-tools to support non standard squashfs filesystems. By the standard, squashfs version 4.0 filesystems are fixed little-endian. There are no big-endian 4.0 filesystems. However, embedded vendors love to mess with standards. So instead of adapting squashfs-tools code, we build a version with `__BYTE_ORDER` swapped which is the `sasquatch-v4be` binary. This binary can handle non-standard squashfs v4 big-endian filesystems.

More information about the big-endian variant can be found at https://github.com/onekey-sec/sasquatch/commit/0c65f00de7c7cc3382302256fddd06ba10ed7b01

Here's a small preview of what vendors do with squashfs. Cells with a star (*) are standard.


type | v2​ | v3​ | v4 | DDWRT v3​ | Broadcom v3​ | ??? v3​ | AVM​ | Netgear​ 
-------|-----|-----|-----|-----------------|---------------------|----------|--------|------------|
magic (LE) | hsqs*​ | hsqs*​ | hsqs*​ | hsqt​ | shsq​  | zlqs​ | Header in BE, but chunks in LE.​ | Let‘s use a non standard LZMA + XZ compression !​ |
magic (BE)​ | sqsh*​ | sqsh*​ | sqsh​ | tqsh​ | qshs​ | sqlz​ | / | / |

File: pkgs/tools/filesystems/sasquatch/default.nix
Comment (qkaiser) [reply pkgs/tools/filesystems/sasquatch/default.nix:15]: The `v4be` indicates the squashfs [filesystem version](https://docs.kernel.org/filesystems/squashfs.html) while the `4.5.1-4` indicates the squashfs-tools version sasquatch is based on. These versions are unrelated.

**Comment by AkechiShiro** [comment]
Hello @smancill I think there was a review that approved the package, I guess running `nixpkgs-review` on this PR and testing the tool and leaving a comment stating how it went could help move the PR forward. 

I don't know however if a rebase is needed to make sure that the current package still build with the latest upstream in Nixpkgs.

**Comment by AkechiShiro** [comment]
@vlaci there might be a dead link as a source for the current version : 
image

I cannot find the repo on GitHub, maybe there is new unblob that is being worked on that doesn't depend on `sasquatch` ?

**Comment by vlaci** [comment]
> @vlaci there might be a dead link as a source for the current version : image
> 
> I cannot find the repo on GitHub, maybe there is new unblob that is being worked on that doesn't depend on `sasquatch` ?
> 
> I only found this repo : https://github.com/onekey-sec/sasquatch
> 
> EDIT :
> 
> * `nixpkgs-review` result :
> 
> ```
> 1 package failed to build:
> sasquatch-v4be
> 
> 14 packages built:
> binwalk binwalk.dist diffoscope diffoscope.dist diffoscope.man python310Packages.binwalk python310Packages.binwalk-full python310Packages.binwalk-full.dist python310Packages.binwalk.dist python311Packages.binwalk python311Packages.binwalk-full python311Packages.binwalk-full.dist python311Packages.binwalk.dist sasquatch
> 
> error: build log of '/nix/store/40nfx0yjl4q7965p05dkqgavd9xjbh7j-sasquatch-v4be-4.5.1-4.drv^*' is not available
> error: build log of '/nix/store/bw99kzndaqc2r8bykzcjc34z91a964j0-sasquatch-v4be-4.5.1-4' is not available
> ```

It must have been a github api error, as the source is there...

**Comment by vlaci** [comment]
IDK if I should move these from `all-packages` accorving to nixpkgs-vet.

**Comment by AkechiShiro** [comment]
I believe there is a typo in the package @vlaci 

```nix
       error: evaluation aborted with the following error message: 'Failed to evaluate sasquatch-4.5.1-4: «unknown-meta»: has an invalid meta attrset:
         - key 'meta.mainprogram' is unrecognized; expected one of:
         ['available', 'badPlatforms', 'branch', 'broken', 'changelog', 'description', 'downloadPage', 'executables', 'homepage', 'hydraPlatforms', 'insecure', 'isBuildPythonPackage', 'isFcitxEngine', 'isGutenprint', 'isHydraChannel', 'isIbusEngine', 'knownVulnerabilities', 'license', 'longDescription', 'mainProgram', 'maintainers', 'maxSilent', 'name', 'outputsToInstall', 'pkgConfigModules', 'platforms', 'position', 'priority', 'schedulingPriority', 'sourceProvenance', 'tag', 'tests', 'timeout', 'unfree', 'unsupported', 'version']
       '
```

`mainprogram` -> `mainProgram`, if you can also solve the conflict and rebase the PR please ?

**Comment by vlaci** [comment]
> I believe there is a typo in the package @vlaci
> 
> ```nix
>        error: evaluation aborted with the following error message: 'Failed to evaluate sasquatch-4.5.1-4: «unknown-meta»: has an invalid meta attrset:
>          - key 'meta.mainprogram' is unrecognized; expected one of:
>          ['available', 'badPlatforms', 'branch', 'broken', 'changelog', 'description', 'downloadPage', 'executables', 'homepage', 'hydraPlatforms', 'insecure', 'isBuildPythonPackage', 'isFcitxEngine', 'isGutenprint', 'isHydraChannel', 'isIbusEngine', 'knownVulnerabilities', 'license', 'longDescription', 'mainProgram', 'maintainers', 'maxSilent', 'name', 'outputsToInstall', 'pkgConfigModules', 'platforms', 'position', 'priority', 'schedulingPriority', 'sourceProvenance', 'tag', 'tests', 'timeout', 'unfree', 'unsupported', 'version']
>        '
> ```
> 
> `mainprogram` -> `mainProgram`, if you can also solve the conflict and rebase the PR please ?

Thanks, this error did not emerge on my Nix version.
Also, I botched up a previous patch, hence the conflict. Adjusted.

File: pkgs/top-level/all-packages.nix
Comment (AkechiShiro) [inline pkgs/top-level/all-packages.nix:12304]: @vlaci could you please maybe move this into `by-name/sa/sasquatch-v4be` as well ?

**Comment by AkechiShiro** [comment]
Tested a build (`nixpkgs-review` seems a bit confused however) with the new changes, seems to be working alright.

Output of `nixpkgs-review` : 
```
8 packages removed:
binwalk (†2.4.2) diffoscope (†277) ghidraninja-ghidra-scripts-unstable (†2020-10-07) python3.11-binwalk (†2.4.2) python3.11-binwalk-full (†2.4.2) python3.12-binwalk (†2.4.2) python3.12-binwalk-full (†2.4.2) sasquatch (†4.4)

Nothing to be built.
```
I ran the build manually.

**Review by AkechiShiro — CHANGES_REQUESTED** [review]
Just waiting for `sasquatch-v4be` to be moved from `top-level/all-packages.nix` into `by-name/sa/sasquatch-v4be`, then I will approve this PR on my side, I can't really test however on nix `darwin`

**Comment by vlaci** [comment]
> Just waiting for `sasquatch-v4be` to be moved from `top-level/all-packages.nix` into `by-name/sa/sasquatch-v4be`, then I will approve this PR on my side, I can't really test however on nix `darwin`

I've created a draft PR for unblob pointing to this branch, so its tests will run for darwin as well: https://github.com/onekey-sec/unblob/pull/946/checks

File: pkgs/top-level/all-packages.nix
Comment (vlaci) [reply pkgs/top-level/all-packages.nix:12304]: Is this how you'd like it?
https://github.com/NixOS/nixpkgs/pull/219169/files#diff-276d44079aa310eda1d4e6bc92d2ffedcf4c0cf9fb5c5f988502809efc60089fR3

Comment (AkechiShiro) [reply pkgs/top-level/all-packages.nix:12304]: Yep thanks a lot for your work on this !

**Comment by AkechiShiro** [comment]
I don't think the `ghidra` failure is related : 
```
1 package failed to build:
ghidra-extensions.ghidraninja-ghidra-scripts

13 packages built:
binwalk binwalk.dist diffoscope diffoscope.dist diffoscope.man python311Packages.binwalk python311Packages.binwalk-full python311Packages.binwalk-full.dist python311Packages.binwalk.dist python312Packages.binwalk-full python312Packages.binwalk-full.dist sasquatch sasquatch-v4be
```

**Review by AkechiShiro — APPROVED** [review]

File: pkgs/by-name/sa/sasquatch/package.nix
Comment (D3vil0p3r) [inline pkgs/by-name/sa/sasquatch/package.nix:15]:
```suggestion
  drv = stdenv.mkDerivation (finalAttrs: {
```

File: pkgs/by-name/sa/sasquatch/package.nix
Comment (D3vil0p3r) [inline pkgs/by-name/sa/sasquatch/package.nix:22]:
```suggestion
      repo = "sasquatch";
      rev = "sasquatch-v${finalAttrs.version}";
```

File: pkgs/by-name/sa/sasquatch/package.nix
Comment (D3vil0p3r) [inline pkgs/by-name/sa/sasquatch/package.nix:62]:
```suggestion
    meta = {
```

File: pkgs/by-name/sa/sasquatch/package.nix
Comment (D3vil0p3r) [inline pkgs/by-name/sa/sasquatch/package.nix:67]:
```suggestion
      license = lib.licenses.gpl2Plus;
      maintainers = with lib.maintainers; [ vlaci ];
      platforms = lib.platforms.unix;
```

File: pkgs/by-name/sa/sasquatch/package.nix
Comment (D3vil0p3r) [inline pkgs/by-name/sa/sasquatch/package.nix:70]:
```suggestion
  });
```

**Review by D3vil0p3r — CHANGES_REQUESTED** [review]
My 2 cents

File: pkgs/by-name/sa/sasquatch/package.nix
Comment (vlaci) [reply pkgs/by-name/sa/sasquatch/package.nix:22]: Nice idea to use fixed-point attribute here, but then shouldn't `hash` also come from an input argument? :thinking:

Comment (D3vil0p3r) [reply pkgs/by-name/sa/sasquatch/package.nix:22]: No. My proposals should make the build working correctly.

Comment (vlaci) [reply pkgs/by-name/sa/sasquatch/package.nix:22]: I understand, but IMO `overrideAttrs (_: { version = "foo"; })` will lead to surprising results, as it will affect src fetching

Comment (vlaci) [reply pkgs/by-name/sa/sasquatch/package.nix:22]: Am I overthinking?

Comment (D3vil0p3r) [reply pkgs/by-name/sa/sasquatch/package.nix:22]: Uhm dunno... feel free to not merge the commit suggestion.

Comment (vlaci) [reply pkgs/by-name/sa/sasquatch/package.nix:22]: Checked that there are many examples in nixpkgs already, so it is fine by me

**Comment by D3vil0p3r** [comment]
@vlaci why do we get those two ofborg-eval errors in check list below?

**Comment by vlaci** [comment]
> @vlaci why do we get those two ofborg-eval errors in check list below?

Silly me, reintroduced a previously fixed bug during last rebase :(

**Review by D3vil0p3r — APPROVED** [review]

**Comment by smancill** [comment]
Result of `nixpkgs-review pr 219169` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>sasquatch</li>
    <li>sasquatch-v4be</li>
  </ul>
</details>

**Pull request merged**
